spec/instrumentation_spec.rb

```diff
@@ -0,0 +1,200 @@
+require 'spec_helper'
+
+RSpec.describe PG::Instrumentation do
+  describe 'Class Methods' do
+    it { should respond_to :instrument }
+    it { should respond_to :patch_methods }
+  end
+
+  let (:tracer) { OpenTracingTestTracer.build }
+
+  let (:host) { 'localhost' }
+  let (:dbname) { 'postgres' }
+  let (:user) { 'test_user' }
+  let (:port) { 5432 }
+
+  let (:conn) { PG::Connection.connect(dbname: dbname, user: user, host: host, port: port)}
+
+  before do
+    PG::Instrumentation.instrument(tracer: tracer)
+
+    ## prevent actual connections
+    allow_any_instance_of(PG::Connection).to receive(:new)
+
+    # mock query_original, since we don't care about the results
+    allow_any_instance_of(PG::Connection).to receive(:initialize_original).and_return(PG::Connection.new)
+    allow_any_instance_of(PG::Connection).to receive(:exec_original).and_return(PG::Result.new)
+    allow_any_instance_of(PG::Connection).to receive(:exec_params_original).and_return(PG::Result.new)
+    allow_any_instance_of(PG::Connection).to receive(:prepare_original).and_return(PG::Result.new)
+    allow_any_instance_of(PG::Connection).to receive(:exec_prepared_original).and_return(PG::Result.new)
+  end
+
+  describe 'fresh start' do
+    it 'cleans up' do
+      conn.exec 'drop table if exists test_pg '
+      conn.exec 'create table test_pg (col1 text, col2 text, col3 text)'
+    end
+  end
+
+  describe :instrument do
+    it "patches the class's exec_params method" do
+      expect(conn).to respond_to(:exec_params)
+      expect(conn).to respond_to(:exec_params_original)
+    end
+
+    it "patches the class's exec method" do
+      expect(conn).to respond_to(:exec)
+      expect(conn).to respond_to(:exec_original)
+    end
+
+    it "patches the class's exec_prepared method" do
+      expect(conn).to respond_to(:exec_prepared)
+      expect(conn).to respond_to(:exec_prepared_original)
+    end
+  end
+
+  describe 'successfully traces query using `exec` ' do
+    it 'calls exec_original when calling query' do
+      expect(conn).to receive(:exec_original)
+
+      conn.exec('SELECT * FROM test_pg')
+    end
+
+    it 'adds a span for a query with tags' do
+      statement = 'SELECT * FROM test_pg '
+      conn.exec(statement)
+
+      expect(tracer.spans.count).to eq 2
+      expect(tracer.spans.last.operation_name).to eq 'SELECT'
+      expected_tags = {
```

Comment:
sry if dup but it feels like gh is dropping my comments.  There's a fair amount of redundancy with these expected tags.  Clone for a single base definition?